cpp/include/partition_manager.hpp

```diff
@@ -0,0 +1,171 @@
+/*
+ * Copyright (c) 2020, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include <raft/comms/comms.hpp>
+#include <raft/handle.hpp>
+
+#include <cassert>
+#include <cmath>
+#include <memory>
+#include <sstream>
+#include <string>
+#include <vector>
+
+namespace cugraph {
+namespace partition_2d {
+template <typename from_t>
+std::string to_string(from_t const& value)
+{
+  std::stringstream ss;
+  ss << value;
+  return ss.str();
+}
+
+// class responsible for creating 2D partition of workers:
+// responsible with finding appropriate P_ROW x P_COL
+// 2D partition and initializing the raft::handle_t communicator
+//
+// (this might be removed; or, it might exist already)
+//
+template <typename size_type = int>
+class partition_manager_t {
+ public:
+  partition_manager_t(raft::handle_t& handle, size_type p_row_size, size_type p_col_size)
+    : handle_(handle), p_row_size_(p_row_size), p_col_size_(p_col_size)
+  {
+    init_communicator();
+  }
+
+  partition_manager_t(raft::handle_t const& handle, size_type p_size) : handle_(handle)
+  {
+    partition2d(p_size);
+    init_communicator();
+  }
+
+  virtual ~partition_manager_t(void) {}
+
+ protected:
+  virtual void partition2d(size_type p_size)
+  {
+    auto sqr = static_cast<size_type>(std::sqrt(p_size));
+
+    // find divisor of p_size
+    // nearest to sqr;
+    //
+    p_row_size_ = nearest_divisor(sqr, p_size);
+    p_col_size_ = p_size / p_row_size_;
```

Comment:
@Iroy30 I remember you asked about finding p_row & p_col if only p is provided. Have you implemented this in python? Then, this code can be redundant.

Reply:
@Iroy30, can you please comment on the above? I'd like to understand the action item for me, here. Thanks in advance.

Reply:
https://github.com/rapidsai/cugraph/pull/1133/files#r488173099 I did implement it in python to be able to shuffle. If it needs to be in c++, I could call that func instead

Reply:
I think it's better to compute this in the python layer as our current software architecture places more control at the python layer, and the C++ layer's main responsibility is in acceleration. This computation is cheap, so no need for acceleration.

Reply:
@aschaffer If this information is computed in python, how/where do you want me to pass it to you.

Reply:
I believe we don't need the partition_manager_t class then and the subcomm_factory_t call could be called with python computed row_size in cython itself if my understanding correct

Reply:
I think init_row_col_comms should take p_row_size and p_col_size as input parameters.

Reply:
@Iroy30, yes, I agree that `partition_manager_t` may not be needed anymore. 

Its raison-d'etre was to fulfill the 1st requirement in https://github.com/rapidsai/cugraph/issues/1065: "Decide P_row & P_column for the given P. We may set P_row and P_column as close as possible to the square root of P by default with a user option to override the default value."

If that is covered in Python code, then I'll just remove this class and keep only ` subcomm_factory_t`. 

The other question that still needs an answer, then, is _where_ to call ` subcomm_factory_t{}` constructor in C++ code?

Reply:
@seunghwak , why take _both_ `p_row_size` *and* `p_col_size`? Shouldn't `p_row_size` be enough? (as it is right now, except it's captured by the constructor). The `rank` info + the `p_row_size` info should be enough to allocate sub-comms, as it is right now. Correct?

Reply:
@aschaffer I think you're right if p_col_size is not used (or can be computed from p_row_size & p) :-)

Reply:
OK, removed `partition_manager_t` class, as discussed. Resolving all conversations related to that class.